frontend/js/src/settings/music-services/details/MusicServices.tsx

```diff
@@ -177,6 +225,113 @@ export default function MusicServices() {
+    evt.preventDefault();
+    try {
+      if (!funkwhaleHostUrl) {
+        throw Error("Funkwhale server URL is required");
+      }
+
+      const response = await fetch(
+        `/settings/music-services/funkwhale/connect/`,
+        {
+          method: "POST",
+          body: JSON.stringify({
+            host_url: funkwhaleHostUrl,
+          }),
+          headers: {
+            "Content-Type": "application/json",
+            Authorization: `Token ${currentUser?.auth_token}`,
+          },
+        }
+      );
+
+      let data;
+      const contentType = response.headers.get("content-type");
+      if (contentType && contentType.includes("application/json")) {
+        data = await response.json();
+      } else {
+        throw Error("Server returned non-JSON response");
+      }
+
+      if (response.ok) {
+        if (data.url) {
+          // Store host URL in session storage before redirect
+          sessionStorage.setItem("funkwhale_host_url", funkwhaleHostUrl);
+          // Redirect to Funkwhale authorization page
+          window.location.href = data.url;
+        } else {
+          throw Error("No authorization URL received from server");
+        }
+      } else if (data?.error) {
+        throw Error(data.error);
+      } else {
+        throw Error(`Server error: ${response.status} ${response.statusText}`);
+      }
+    } catch (error) {
+      console.error("Funkwhale connection error:", error);
+      toast.error(
+        <ToastMsg
+          title="Failed to connect to Funkwhale"
+          message={error.toString()}
+        />
+      );
+    }
+  };
+
+  React.useEffect(() => {
+    const params = new URLSearchParams(window.location.search);
+    const funkwhaleError = params.get("error");
+    const funkwhaleSuccess = params.get("success");
+
+    if (funkwhaleSuccess === "Successfully connected to Funkwhale") {
+      // Get host URL from session storage
+      const hostUrl = sessionStorage.getItem("funkwhale_host_url");
+      if (hostUrl) {
+        setFunkwhaleHostUrl(hostUrl);
+        setConnectedFunkwhaleServers((prev) => [...prev, hostUrl]);
+        sessionStorage.removeItem("funkwhale_host_url"); // Clean up
+      }
+      toast.success(
+        <ToastMsg
+          title="Success"
+          message="Successfully connected to Funkwhale!"
+        />
+      );
+      setPermissions((prev) => ({ ...prev, funkwhale: "listen" }));
+      // Clear the query parameters from the URL
+      window.history.replaceState(
+        {},
+        document.title,
+        window.location.pathname + window.location.hash
+      );
+    } else if (funkwhaleError) {
+      toast.error(
+        <ToastMsg
+          title="Funkwhale Connection Error"
+          message={decodeURIComponent(funkwhaleError)}
+        />
+      );
+      // Clear the query parameters from the URL
+      window.history.replaceState(
+        {},
+        document.title,
+        window.location.pathname + window.location.hash
+      );
```

Comment:
This code is repeated in both if/else code blocks, so it can simply be deduplicated and moved to after the if/else block to be run in both cases.